### DISCUSSION TITLE
Handling next | Redirect only sometimes works

### DISCUSSION THREAD

**Discussion by sai-sy** [maxcountryman/flask-login#653] [Q&A] [answered]
I've been making a web app over the last couple weeks and it worked fine, I deployed it for production testing and it was great. Then I switched to safari on my phone it wouldn't redirect out of the login page. After quick googling it seemed to be about using the next value which most tutorials don't use but is in the documentation. However no matter the method I use to implement this, it seems to not work on some phone somewhere or some browser some where. The same issue maintains where once you click submit, it does not redirect anywhere.

### HTML
```html
<body>
    {%extends 'base.html'%}
    {% block content %}
    <div class="shadow p-3 mb-5 bg-body rounded">
    <h3 align="center">Login</h3>
    <form method="POST" action="{{url_for('auth.login')}}">
        {{ form.hidden_tag() }}
        {{ form.email.label(class='form-label')}}
        {{ form.email(class='form-control')}}
        <br/>
        {{ form.hidden_tag() }}
        {{ form.password.label(class='form-label')}}
        {{ form.password(class='form-control')}}
        <br/>
        
        <input type="hidden" name="next" value="{{ request.args.get('next', '') }}"/>
        {{ form.submit(class='btn btn-danger')}}
    </form>
    </div>
    {% endblock %}
</body>
</html>
```
### I want to redirect to views.home which goes to shift_route.shift_add
# Method One
`auth.py`
```python
@auth.route('/login', methods=['GET', 'POST'])
def login():
    #setting up the form
    
    if form.validate_on_submit():
        email = form.email.data
        password = form.password.data

        user = Users.query.filter_by(email=email).first()
        if user:
            if check_password_hash(user.password_hash, password):
                flash('Logged in successfully!', category='success')
                login_user(user, remember=True)
                form.email.data = ''
                form.password.data = ''
                
                # Method One
                return redirect(url_for('views.home'))
            else:
                form.email.data = ''
                form.password.data = ''
                flash('Incorrect password, ty again.', category='error')
        else:
            form.email.data = ''
            form.password.data = ''
            flash('Email does not exist', category='error')

        form.email.data = ''
        form.password.data = ''
    return render_template('/user/login.html', form=form)
```
### Firefox
In firefox this returns me to /views.none
### Safari
In safari it just keeps me at the login page and clears the password field, this is the output in the logs
```
00.000.000.000 - - [02/Apr/2022:01:10:01 +0000] "POST /login HTTP/1.1" 200 1947 "https://websitename.com/login" "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1"
00.000.000.000 - - [02/Apr/2022:01:10:01 +0000] "GET /static/stylesheet.css HTTP/1.1" 403 134 "https://websitename.com/login" "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1"
00.000.000.000 - - [02/Apr/2022:01:10:01 +0000] "GET /dist/js/BsMultiSelect.min.js HTTP/1.1" 404 1565 "https://websitename.com/login" "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1"
```
### Chrome
Chrome sends me to https://websitename.com/login?next=%2Fshift_add and keeps me there after submit
```
00.000.000.000 - - [02/Apr/2022:01:12:05 +0000] "POST /login?next=%2Fshift_add HTTP/1.1" 200 1958 "https://websitename.com/login?next=%2Fshift_add" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36"
00.000.000.000 - - [02/Apr/2022:01:12:05 +0000] "GET /static/stylesheet.css HTTP/1.1" 403 196 "https://websitename.com/login?next=%2Fshift_add" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36"
00.000.000.000 - - [02/Apr/2022:01:12:05 +0000] "GET /dist/js/BsMultiSelect.min.js HTTP/1.1" 404 1565 "https://websitename.com/login?next=%2Fshift_add" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.60 Safari/537.36"
```
# Method Two
`auth.py`
```python
                next = request.args.get('next')
                current_app.logger.info(next)
                if not is_safe_url(next):
                    return abort(400)
                else:
                    return redirect(next or url_for('views.home'))
```
### Firefox
Firefox from /login after submit takes me to /views.None
Firefox from  /login?next=%2Fshift_add  after submit takes me to /views./shift_add

### Safari
Safari now opens to https://websitename.com/login?next=%2Fshift_add and then just keeps me at this link

### Chrome
Chrome opens me to https://websitename.com/login?next=%2Fshift_add and then after submit just keeps me here

# Method Three
This only changes the order of the redirect
`auth.py`
```python
                next_url = request.form.get("next")
                if next_url:
                    return redirect(next_url)
                return redirect(url_for('views.home') or next)
```

### Firefox
Firefox from /login after submit takes me to /views.None
Firefox from  /login?next=%2Fshift_add  after submit takes me to /views./shift_add

### Safari
Safari now opens to https://websitename.com/login?next=%2Fshift_add and then just keeps me at this link

### Chrome 
Chrome now opens to https://websitename.com/login?next=%2Fshift_add and then just keeps me at this link

# Method Four
`auth.py`
```python
return redirect('views.' + str(request.args.get("next")) or url_for("views.home"))
```
Same results across the board

# Question
Maybe one of these methods should be working, but I can't pin point which is the best way to implement it, I don't get where views./shift_add is coming from, Even if not solutions, guidance on which method I should be using or how my html should look would be great.

**Comment by alanhamlett** [comment]
Your code isn't reaching your redirect so `return redirect(url_for('views.home'))` never runs, because either `if user` or `if check_password_hash(user.password_hash, password)` is evaluating to `False`. You need either to return redirect or always set an error message on the page when `request.method == "POST"`. Look at the WTForms library for examples.

**Reply by sai-sy** [reply]
Sorry I should've been more clear apparently I forgot to include the full route, I've added it to the first python block. I have handling for if user doesn't pass or if check_password doesn't pass. And it doesn't make sense on why it would pass on Firefox but not other browsers right unless I'm mishandling something.

**Reply by sai-sy** [reply]
The error messages get displayed in the {%extends 'base.html'%} block. I think maybe my form might not be doing the right action thought.

**Reply by sai-sy** [reply]
You are right though that it might not be the redirect itself that's the issue but if so then which one of the methods should I use. One is just a regular redirect. Two is just the way flask login docs say to use it, and the final one is off of stack overflow.

**Comment by sai-sy** [comment]
I found in the DOM that I get this error
![dom error](https://user-images.githubusercontent.com/74868837/161381311-45d1a5d2-2969-4a05-b9d0-a5478f3be601.png)
Because of this I updated my html to look like
```html
<body>
    {%extends 'base.html'%}
    {% block content %}
    <div class="shadow p-3 mb-5 bg-body rounded">
    <h3 align="center">Login</h3>
    <form method="POST">
        {{ form.hidden_tag() }}
        
        {{ form.email.label(class='form-label')}}
        {{ form.email(class='form-control')}}
        <br/>
        {{ form.password.label(class='form-label')}}
        {{ form.password(class='form-control')}}
        <br/>
        {{ form.submit(class='btn btn-danger')}}
    </form>
    </div>
    {% endblock %}
</body>
</html>
``` 
however the same error persists. Online I can only find people who's error is because of two forms but here I of course only have one. This question has maybe moved a bit towards Flask-WTF however I still am curious on which implementation of the next arg should I use for the login.

**Comment by sai-sy** [comment, marked as answer]
The answer seems was a couple things, I needed to set a SERVER_NAME env variable for wtforms csrf token to work. Everything else fell into place after I fixed that.
